File: frontend/src/components/Smtp/controller/Smtp.Service.ts
/// <reference path="../../app/App.Resource.ts" />
/// <reference path="../../Smtp/model/Smtp.Model.ts" />

/*******************************************************************************
 * Copyright (c) 2016 Eviarc GmbH.
 * All rights reserved.  
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Eviarc GmbH and its suppliers, if any.  
 * The intellectual and technical concepts contained
 * herein are proprietary to Eviarc GmbH,
 * and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Eviarc GmbH.
 *******************************************************************************/
"use strict";

const SmtpServiceId: string = "SmtpService";

class SmtpService {

    private $inject = [toasterId, $translateId, $rootScopeId, SmtpResourceId, UserResourceId, $qId];

    smtpResource;
    userResource;
    q;
    currentSmtp: Smtp;

    rootScope;
    translate;
    toaster;
    currentPasswordLength: number = 10;

    constructor(toaster, $translate, $rootScope, SmtpResource, UserResource, $q) {
        this.toaster = toaster;
        this.rootScope = $rootScope;
        this.translate = $translate;
        this.q = $q;
        this.smtpResource = SmtpResource.resource;
        this.userResource = UserResource.resource;
        this.getSMtp();
    }

    test() {
        let self = this;
        this.save().then(function () {
            self.smtpResource.testSmtp({ id: self.currentSmtp.id, smtpKey: self.rootScope.globals.user.smtpKey }).$promise.then(function () {
                self.toaster.pop("success", "", self.translate.instant("SETTING_TOAST_EMAIL_MANAGEMENT_CONNECTION_TEST"));
            }, function () {
                self.toaster.pop("error", "", self.translate.instant("SETTING_TOAST_EMAIL_MANAGEMENT_CONNECTION_TEST_ERROR"));
            });
        });

    }

    getSMtp() {
        let self = this;
        this.smtpResource.getByUserId({ id: this.rootScope.globals.user.id }).$promise.then((data) => {
            self.currentSmtp = data;
            self.currentSmtp.stringPassword = "";
            for (let i = 0; i < self.currentPasswordLength; i++) {
                self.currentSmtp.stringPassword += "x";
            }
        });
    }
    save(): IPromise<any> {
        let defer = this.q.defer();
        this.currentSmtp.user = this.rootScope.globals.user;
        this.currentPasswordLength = this.currentSmtp.stringPassword !== null ? this.currentSmtp.stringPassword.length : this.currentPasswordLength;
        this.currentSmtp.password = this.currentSmtp.stringPassword !== null ? btoa(this.currentSmtp.stringPassword) : null;
        let self = this;
        this.smtpResource.createSmtp({ smtpKey: self.rootScope.globals.user.smtpKey }, this.currentSmtp).$promise.then(function (data) {
            self.currentSmtp = data;
            self.currentSmtp.stringPassword = "";
            for (let i = 0; i < self.currentPasswordLength; i++) {
                self.currentSmtp.stringPassword += "x";
            }
            self.toaster.pop("success", "", self.translate.instant("SETTING_TOAST_EMAIL_MANAGEMENT_CONNECTION_SAVE"));
            defer.resolve(self.currentSmtp);
        }, function (error) {
            self.toaster.pop("error", "", self.translate.instant("SETTING_TOAST_EMAIL_MANAGEMENT_CONNECTION_SAVE_ERROR"));
            defer.reject(error);
        });
        return defer.promise;
    }
}

angular.module(moduleSmtpService, [ngResourceId]).service(SmtpServiceId, SmtpService);

File: src/main/java/dash/smtpmanagement/business/SmtpService.java
/*******************************************************************************
 * Copyright (c) 2016 Eviarc GmbH.
 * All rights reserved.  
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Eviarc GmbH and its suppliers, if any.  
 * The intellectual and technical concepts contained
 * herein are proprietary to Eviarc GmbH,
 * and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from Eviarc GmbH.
 *******************************************************************************/

package dash.smtpmanagement.business;

import javax.mail.Session;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;

import dash.common.EncryptionWrapper;
import dash.common.Encryptor;
import dash.exceptions.NotFoundException;
import dash.notificationmanagement.business.NotificationUtil;
import dash.notificationmanagement.domain.Notification;
import dash.security.business.ISecurityService;
import dash.security.jwt.domain.UserContext;
import dash.smtpmanagement.domain.Smtp;
import dash.smtpmanagement.domain.SmtpEncryptionType;

@Service
public class SmtpService implements ISmtpService {

	@Autowired
	private SmtpRepository smtpRepository;

	private ISecurityService securityService;

	public SmtpService(ISecurityService securtiyService) {
		this.securityService = securtiyService;
	}

	@Override
	public Smtp test(final Long smtpId) throws Exception {

		Smtp smtp = this.smtpRepository.findOne(smtpId);
		if (smtp == null)
			throw new NotFoundException("There is no SMTP-Server Configuration for specified id.");

		smtp.setVerified(false);
		// SMTP-Access
		String username = smtp.getUsername();
		String emailFrom = smtp.getEmail();
		String senderName = smtp.getSender();
		SmtpEncryptionType smtpEncryptionType = smtp.getEncryption();
		String password;

		// SMTP-Server
		String host = smtp.getHost();
		int port = smtp.getPort();

		// SMTP-Session
		Session session;

		Notification notification;

		password = SmtpUtil.decryptPasswordForSmtp(smtp);
		session = SmtpUtil.createSessionWithAuthentication(host, port, smtpEncryptionType, username, password);
		notification = NotificationUtil.createTestNotification(smtp.getEmail(), smtp.getSender());
		Smtp tempSmtp = null;
		try {
			if (smtpEncryptionType == SmtpEncryptionType.PLAIN) {
				NotificationUtil.plainTransport(session, notification, emailFrom, senderName, host, port, username,
						password);
			} else if (smtpEncryptionType == SmtpEncryptionType.TLS || smtpEncryptionType == SmtpEncryptionType.SSL
					|| smtpEncryptionType == SmtpEncryptionType.STARTTLS) {
				NotificationUtil.secureTransport(session, notification, emailFrom, senderName, host, port, username,
						password);
			}
			smtp.setVerified(true);

		} finally {

			tempSmtp = save(smtp, null);
		}
		return tempSmtp;
	}

	@Override
	public Smtp save(final Smtp smtp, String smtpKey) throws Exception {

		if (smtp != null && smtp.getId() != null
				&& (smtp.getPassword() == null || new String(smtp.getPassword(), "UTF-8") == "")) {
			Smtp tempSmpt = smtpRepository.findOne(smtp.getId());
			smtp.setPassword(tempSmpt.getPassword());
			smtp.setSalt(tempSmpt.getSalt());
			smtp.setIv(tempSmpt.getIv());
		} else if (smtp != null && smtp.isDecrypted() == true && smtp.getPassword() != null
				&& new String(smtp.getPassword(), "UTF-8") != "") {
			if (smtpKey == null) {
				Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
				UserContext userContext = (UserContext) authentication.getPrincipal();
				smtpKey = userContext.getSmtpKey();
			}

			EncryptionWrapper encryptionWrapper = Encryptor.encrypt(smtp.getPassword(), smtpKey);
			smtp.setPassword(encryptionWrapper.getCiphertext());
			smtp.setSalt(encryptionWrapper.getSalt());
			smtp.setIv(encryptionWrapper.getIv());
		}
		return smtpRepository.save(smtp);
	}

	@Override
	public Smtp findByUserId(final long userId) {
		try {
			return smtpRepository.findByUserId(userId);
		} catch (NotFoundException e) {
			return null;
		}
	}

	@Override
	public Smtp findByAuthenticatedUser() throws NotFoundException {
		return smtpRepository.findByUserUsername(this.securityService.getAuthenticatedUser());
	}

}
